Tests/SwiftLexicalLookupTest/Assertions.swift

```diff
@@ -17,11 +17,65 @@ import SwiftSyntax
 import XCTest
 import _SwiftSyntaxTestSupport
 
-/// `methodUnderTest` is called with the token at every position marker in the keys of `expected`. It then asserts that the positions of the syntax nodes returned by `methodUnderTest` are the values in `expected`.
+/// Used to define result type expectectations for given markers.
+enum MarkerExpectation {
+  /// Specifies a separate type for each result marker.
+  case distinct([String: SyntaxProtocol.Type])
+  /// Specifies a common type for all results
+  /// apart from the ones defined explicitly in `except`.
+  case all(SyntaxProtocol.Type, except: [String: SyntaxProtocol.Type] = [:])
+  /// Does not assert result types.
+  case none
+
+  /// Assert `actual` result labeled with `marker`
+  /// according to the rules represented by this expectation.
+  fileprivate func assertMarkerType(marker: String, actual: SyntaxProtocol) {
+    switch self {
+    case .all(let expectedType, except: let dictionary):
+      assertMarkerType(marker: marker, actual: actual, expectedType: dictionary[marker] ?? expectedType)
+    case .distinct(let dictionary):
+      if let expectedType = dictionary[marker] {
+        assertMarkerType(marker: marker, actual: actual, expectedType: expectedType)
+      } else {
+        XCTFail("For result \(marker), could not find type expectation")
+      }
+    case .none:
+      break
+    }
+  }
+
+  /// Assert whether `actual` type matches `expectedType`.
+  private func assertMarkerType(marker: String, actual: SyntaxProtocol, expectedType: SyntaxProtocol.Type) {
+    XCTAssert(
+      actual.is(expectedType),
+      "For result \(marker), expected type \(expectedType) doesn't match the actual type \(actual.syntaxNodeType)"
+    )
+  }
+}
+
+/// Used to define
+enum ResultExpectation {
+  case fromScope(ScopeSyntax.Type, expectedNames: [String])
+  case fromFileScope(expectedNames: [String])
+
+  var expectedNames: [String] {
+    switch self {
+    case .fromScope(_, let expectedNames):
+      expectedNames
+    case .fromFileScope(expectedNames: let expectedNames):
+      expectedNames
+    }
+  }
+}
+
+/// `methodUnderTest` is called with the token at every position marker in the keys of `expected`.
+/// It then asserts that the positions of the syntax nodes returned by `methodUnderTest` are the values in `expected`.
+/// It also checks whether result types match rules specified in `expectedResultTypes`.
 func assertLexicalScopeQuery(
   source: String,
-  methodUnderTest: (TokenSyntax) -> ([SyntaxProtocol?]),
-  expected: [String: [String?]]
+  methodUnderTest: (String, TokenSyntax) -> ([SyntaxProtocol?]),
```

Comment:
Do you think that adding argument labels to the closure would help clarify especially what the `String` parameter is?

```suggestion
  methodUnderTest: (_ marker: String, _ tokenAtMarker: TokenSyntax) -> ([SyntaxProtocol?]),
```

Reply:
I think that’s a good idea. It was clearer before, when only passing tokens as arguments, that’s why I didn’t add labels earlier. I added the labels now.